### ISSUE TITLE
Invalid trimming suppression in AIJsonUtilities

### ISSUE THREAD

**Issue by eerhardt** [dotnet/extensions#5629]
We have the following trimming suppression:

https://github.com/dotnet/extensions/blob/6487428046a2d7a5c8bb6034e97c38d6e0b7affa/src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Schema.cs#L193-L198

This is an incorrect way to address trimming warnings. The intention of the warnings is so developers now upfront what does and what doesn't work when the build and publish a trimmed or native AOT'd app. Suppressing warnings like this minimizes the principle of "if you don't get warnings during publish, your app is guaranteed to work the same as before you published".

We should remove this suppression and either make the underlying code trim-compatible, or we should annotate this method as `RequiresUnreferencedCode`.

cc @eiriktsarpalis

**Comment by eiriktsarpalis** [comment]
It's something we debated when the polyfill got introduced. Ultimately, the desire to avoid viral annotations prevailed. cc @stephentoub

**Comment by eiriktsarpalis** [comment]
> In this case, you get an exception the moment you hit this in your app that tells you exactly what the problem is and how to fix it.

The runtime exceptions have actually been eliminated with some clever use of reflection Eric proposed. The issue now only manifests as missing custom attributes (nullability annotations, description attributes). The workaround to disable `IlcTrimMetadata` remains though.

**Comment by stephentoub** [comment]
> The issue now only manifests as missing custom attributes (nullability annotations, description attributes).

Then even more so I'm happier suppressing here, since the impact is minimal but the cost of not suppressing like this is high.

That said, it sounds like the justification is now wrong and should be fixed.

**Comment by eerhardt** [comment]
> But not suppressing harms the experience for everyone else, by forcing APIs to be surfaced in particularly non-ergonomic ways.

The same APIs can be surfaced in the same way. You would just need to flow the `[RequiresUnreferencedCode]` annotations up to all callers of this method for the non-.NET 9 build.

> The runtime exceptions have actually been eliminated

This is part of why the suppressions are bad. The static analysis is disabled, so it can't help you anymore.

>  The workaround to disable IlcTrimMetadata remains though.

@agocke @MichalStrehovsky - I don't see any documentation on that switch on learn.microsoft.com.

**Comment by stephentoub** [comment]
> The same APIs can be surfaced in the same way. You would just need to flow the [RequiresUnreferencedCode] annotations up to all callers of this method for the non-.NET 9 build.

Yes, and then to allow the functionality to be used correctly, you need to expose other APIs that complicate the surface area.

**Comment by eerhardt** [comment]
> Yes, and then to allow the functionality to be used, you need to expose other APIs that complicate the surface area.

How does suppressing the warning change this? Why don't we need these other APIs now?

> How would [RequiresUnreferencedCode] help with that?

It turns this case from a possible false-negative into a possible false-postive, which is more desirable. It is more desirable to get a warning, when in fact it is fine, than it is to not get a warning for something that is going to break.

**Comment by stephentoub** [comment]
>> Yes, and then to allow the functionality to be used, you need to expose other APIs that complicate the surface area.

> How does suppressing the warning change this? Why don't we need these other APIs now?

If we propagate the [RUC] all the way up to public surface area, then for example this needs to show up on AIFunctionFactory.Create, which optionally accepts an AIFunctionFactoryCreateOptions, which optionally contains a JsonSerializerOptions. If that Create is [RUC], then to make the whole factory usable with trimming / NativeAOT, we can't expose it like that, and instead need some other API that forces you to provide a JSO when creating an AIFunctionFactoryCreateOptions and forces you to supply an AIFunctionFactoryCreateOptions to Create.

**Comment by eerhardt** [comment]
Here's all the public API that would need to be annotated https://github.com/dotnet/extensions/commit/2b17dffd82efbb70b9caa54aff430b00ccbd0a71

**Comment by eerhardt** [comment]
> If that Create is [RUC], then to make the whole factory usable with trimming / NativeAOT, we can't expose it like that

Note it is only [RUC] on pre-.NET 9 builds, where you aren't using System.Text.Json v9.

**Comment by eerhardt** [comment]
AFAICT - these are the 2 places that are problematic:

![Image](https://github.com/user-attachments/assets/f4f29e4c-db0f-4501-8837-3614d8e09291)

https://github.com/dotnet/extensions/blob/6487428046a2d7a5c8bb6034e97c38d6e0b7affa/src/Shared/JsonSchemaExporter/JsonSchemaExporter.cs#L261-L276

Doing a visual inspection, I don't see any runtime error that will happen if either the constructors (1st case) or the property/field metadata (2nd case) are trimmed. So what will happen is that when you try using these APIs in a trimmed .NET 8 app, you can get a different JSON schema depending on what reflection info was available. If no one else is reflecting on the constructor or properties, the metadata won't be there, and the JSON schema will be missing that information. The property attributes appear to be used for nullability and description info, as @eiriktsarpalis points out above. The constructor is used to know which ctor parameter maps to which property.

This is the bad scenario - your app "kinda works" like it did before trimming. But it works in a different way now, because the JSON schema is different. There is no way for you to know that your app is going to behave differently other than by exercising every line after trimming.

> The problem with these annotations is that they're not actionable. There is no alternative API users can call that gets rid of it. They have to either propagate the annotation on their code or suppress it.

The action would be to target `net9.0` if you want to trim/native AOT your app. We can add that to the RUC message. The above commit was just to show which APIs would need to be annotated.

**Comment by eiriktsarpalis** [comment]
Related to https://github.com/eiriktsarpalis/stj-schema-mapper/issues/7

**Comment by MichalStrehovsky** [comment]
We guarantee to users that either there will be trimming warnings, or the app will behave the same before/after trimming. If you suppress a warning and this doesn't hold, the suppression is simply a bug. It would be better to do _nothing_ (no suppression, just turn off the trimming analyzer on the AIJsonUtilities project).

We don't ask users to re-test the entire app after trimming. By suppressing a warning you put them into a position where they would need to re-test entire app every time they publish. Minor changes in completely unrelated parts of the codebase could make it so that this throws or doesn't throw. Whether something is visible from reflection or isn't can be affected by butterfly effects in unrelated parts of the app.

> [@agocke](https://github.com/agocke) [@MichalStrehovsky](https://github.com/MichalStrehovsky) - I don't see any documentation on that switch on learn.microsoft.com.

That's because it's an undocumented, and unsupported switch. We only have it due to testing in the dotnet/runtime repo and trying to phase it out (https://github.com/dotnet/runtime/pull/91774). There's no intention to ship it. It doesn't "solve" trimming.

**Comment by stephentoub** [comment]
@eiriktsarpalis, this can be closed now, right?

**Comment by eiriktsarpalis** [comment]
Yep, this got solved once we moved to STJv10.
